### ISSUE TITLE
Strange behavior while change variable s from torch.Tensor to torch.nn.Paramter

### ISSUE THREAD

**Issue by keshawnhsieh** [moskomule/l0.pytorch#3]
Hi, moskomule. I have observed a strange behavior when I was attempting to add a HISTOGRAM watch in Tensorboard w.r.t masked weight. I replaced code below

    def _get_mask(self):
        if self.training:
            self.uniform.uniform_()
            u = Variable(self.uniform)
            s = torch.sigmoid((torch.log(u) - torch.log(1 - u) + self.loc) / self.temp)
            s = s * (self.zeta - self.gamma) + self.gamma
            penalty = torch.sigmoid(self.loc - self.temp * self.gamma_zeta_ratio).sum()
        else:
            s = torch.sigmoid(self.loc) * (self.zeta - self.gamma) + self.gamma
            penalty = 0
        return hard_sigmoid(s), penalty

with

    ......
    self.mask = nn.Parameter(torch.zeros(self._size), requires_grad=False)
    def _get_mask(self):
        if self.training:
            self.uniform.uniform_()
            u = Variable(self.uniform)
            s = torch.sigmoid((torch.log(u) - torch.log(1 - u) + self.loc) / self.temp)
            s = s * (self.zeta - self.gamma) + self.gamma
            penalty = torch.sigmoid(self.loc - self.temp * self.gamma_zeta_ratio).sum()
        else:
            s = torch.sigmoid(self.loc) * (self.zeta - self.gamma) + self.gamma
            penalty = 0
        self.mask.data = hard_sigmoid(s)
        return self.mask, penalty

Here I introduced variable named self.mask with type of nn.Parameter in order to easily export masked weight to Tensorboard while training.
But after such modification, I noticed that the loss began to increase after 3 or 4 epochs and the training/validation accuracy began to drop quickly at the same time. 
I was confused why such minor modification could cause such strange behavior.

**Comment by moskomule** [comment]
Hi, What about recording `mask` in a buffer and report buffers as well as parameters?

**Comment by keshawnhsieh** [comment]
Thanks, it works. But I still could not figure out why my previous change could cause such strange problem. I will dig into it when I have time.

**Comment by moskomule** [comment]
Great! But still mysterious for me either. Let me know if you find something useful.

**Comment by keshawnhsieh** [comment]
One more question, I was wondering why the hard concrete distribution gate is implemented like this during training.
```
u = Variable(self.uniform)
s = torch.sigmoid((torch.log(u) - torch.log(1 - u) + self.loc) / self.temp)
s = s * (self.zeta - self.gamma) + self.gamma
s = hard_sigmoid(s)
```
However, it is implemented like this during test.
```
s = torch.sigmoid(self.loc) * (self.zeta - self.gamma) + self.gamma
s = hard_sigmoid(s)
```
The reason why I was curious about this is that I add these two distributions into Tensorboard and found that the distribution of s used in testing is less sparse than in training. I posted two pictures showing their distributions below. Left is training and right is testing
![image](https://user-images.githubusercontent.com/29997158/51308158-730cdb00-1a7c-11e9-87fe-16a9a8aa18cc.png)
![image](https://user-images.githubusercontent.com/29997158/51308172-7acc7f80-1a7c-11e9-8852-23caa4423335.png)


BTW, my goal is to prune neural networks trained with L0 norm according to the sparsity of gate s. So the more sparse of the gate s is, the better result I will  hopefully get.

**Comment by moskomule** [comment]
I implemented this far before, and I cannot remember why. I think I tried to implement as written in the paper.
